[AMDGPU][Scheduler] Refactor ArchVGPR rematerialization during scheduling

lucas-rami opened the pull request:
AMDGPU scheduler's `PreRARematStage` attempts to increase function occupancy w.r.t. ArchVGPR usage by rematerializing trivial ArchVGPR-defining instruction next to their single use. It first collects all eligible trivially rematerializable instructions in the function, then sinks them one-by-one while recomputing occupancy in all affected regions each time to determine if and when it has managed to increase overall occupancy. If it does, changes are committed to the scheduler's state; otherwise modifications to the IR are reverted and the scheduling stage gives up.

In both cases, this scheduling stage currently involves repeated queries for up-to-date occupancy estimates and some state copying to enable reversal of sinking decisions when occupancy is revealed not to increase. The current implementation also does not accurately track register pressure changes in all regions affected by sinking decisions.

This commit refactors this scheduling stage, improving RP tracking and splitting the stage into two distinct steps to avoid repeated occupancy queries and IR/state rollbacks.

- Analysis and collection (`canIncreaseOccupancyOrReduceSpill`). The number of ArchVGPRs to save to reduce spilling or increase function occupancy by 1 (when there is no spilling) is computed. Then, instructions eligible for rematerialization are collected, stopping as soon as enough have been identified to be able to achieve our goal (according to slightly optimistic heuristics). If there aren't enough of such instructions, the scheduling stage stops here.
- Rematerialization (`rematerialize`). Instructions collected in the first step are rematerialized one-by-one. Now we are able to directly update the scheduler's state since we have already done the occupancy analysis and know we won't have to rollback any state. Register pressures for impacted regions are recomputed only once, as opposed to at every sinking decision.

In the case where the stage attempted to increase occupancy, and if both rematerializations alone and rescheduling after were unable to improve occupancy, then all rematerializations are rollbacked.

This replaces #118722, which seems to be stuck in an infinite processing loop.

lucas-rami committed:
Integrate #124327

lucas-rami committed:
Fix spurious auto-formats and typo

The typo in GCNSubtarget::getMaxNumArchVGPRs made the function return
the wrong value in all cases. The returned value is now correct; this
causes some unit tests to break in largely cosmetic ways, since that
value is used to calculate excess RP in the remat stage.

lucas-rami committed:
Fix broken unit test

lucas-rami committed:
Merge branch 'main' into pre-ra-remat-refactor

lucas-rami committed:
Reintroduce missing tests and fix existing one

lucas-rami commented on llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir:
```diff
@@ -6047,9 +7328,9 @@ body:             |
 
   bb.4:
 
+    S_NOP 0, implicit %52, implicit %62
```
In a few cases here and below I moved this instruction to the top of the region to allow rematerializations to increase occupancy. Otherwise the instruction defining %60 is rematerialized at the beginning of the region which doesn't decrease maximum RP in the region, even though my optimistic heuristics estimate that it should have been enough. Ideally, I think the scheduler should be able to reorder instructions in such cases, but it does not. Alternatively, we could make the stage perform more rematerializations when it realizes pre-rescheduling that it did not achieve the expected goal.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,249 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+      ST.getBaseMaxNumVGPRs(MF.getFunction(),
+                            {ST.getMinNumArchVGPRs(OccBounds.second),
+                             ST.getMaxNumArchVGPRs(OccBounds.first)},
+                            false);
```
Does the maxNumArchVGPRs calculation take into account min-waves-per-eu?

lucas-rami replied:
It didn't before but does now. Since the default handling of this attribute yields unexpected results I handle it directly in the stage instead of relying on the subtarget (which imo should be fixed). Basically I want to allow the user to request lower occupancies than the bounds induced by workgroup sizes.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,249 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    GCNRegPressure &RP = DAG.Pressure[I];
+    unsigned NumVGPRs = RP.getArchVGPRNum();
+    unsigned ExcessRP = 0;
+    if (NumVGPRs > MaxVGPRsNoSpill) {
```
We can also spill if AGPR + VGPR > getMaxNumVGPRs (for the minimum number of waves). For spilling of this type, it may not be enough to reduce just ArchVGPR pressure

lucas-rami replied:
Thanks for the catch, I added `PreRARematStage::hasExcessVGPRs` to perform this calculation correctly. See test `reduce_arch_and_acc_vgrp_spill` for an example of how the behavior differs depending on whether the target has a unified RF.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,249 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    GCNRegPressure &RP = DAG.Pressure[I];
+    unsigned NumVGPRs = RP.getArchVGPRNum();
+    unsigned ExcessRP = 0;
+    if (NumVGPRs > MaxVGPRsNoSpill) {
```
If we first encounter a region with occupancy min occupancy and no spilling before encountering a region with spilling, will we try to reduce RP in the first region to achieve increased occupancy?

lucas-rami replied:
Not currently. I err on the conservative side (i.e. do not remat too much) for now, so if any region has spilling w.r.t. to the minimum achievable/requested occupancy I will just try to reduce/eliminate that spilling.

jrbyrnes replied:
Yeah, that makes sense.

But (I could be misreading here) it looks like the IncreaseOccupancy flag is only set to false when we first encounter a region with spilling. So it looks like any region before thhat first spilling region will be inserted into OptRegions so long as the region is at the minimum occupancy?

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,249 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+        OptRegions.clear();
+      } else if (NumVGPRs > MaxVGPRsIncOcc) {
+        // Occupancy is VGPR-limited.
+        ExcessRP = NumVGPRs - MaxVGPRsIncOcc;
```
Occupancy is based on AGPR + VGPR , the calculation for ArchVGPR reduction needed should take into account current ArchVGPR + AGPR usage (i.e. it may be the case that ArchVGPR usage alone is already sufficient for increasing occupancy).

lucas-rami replied:
Thanks for the catch, should be handled correctly by `PreRARematStage::hasExcessVGPRs` as well now.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1989,45 +2040,115 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+void PreRARematStage::finalizeGCNSchedStage() {
+  // We consider that reducing spilling is always beneficial so we never
+  // rollback rematerializations in such cases. It's also possible that
+  // rescheduling lowers occupancy over the one achived just through remats, in
```
Typo: achieved, lowers occupancy -> increases occupancy

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1989,45 +2040,115 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+  // rollback rematerializations in such cases. It's also possible that
+  // rescheduling lowers occupancy over the one achived just through remats, in
+  // which case we do not want to rollback either.
+  if (!IncreaseOccupancy || AchievedOcc == TargetOcc)
```
should AchievedOcc be updated after scheduling the regions? Otherwise, we aren't actually checking if scheduling has improved occupancy when remat alone wasn't sufficient

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1989,45 +2040,115 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+    // Re-rematerialize MI at the end of its original region. Note that it may
+    // not be rematerialized exactly in the same position as originally within
+    // the region, but it should not matter much. Since we are only
+    // rematerializing instructions that do not have any virtual reg uses, we
```
we remat instructions with virtual reg uses so the comment is incorrect. However, all the uses will be live out to the region so the logic is still valid.

lucas-rami committed:
Address feedback

- Properly account for AGPRs in excess RP calculations.
- Account for SGPR spilling when deciding whether to try to increase
  occupancy.
- Don't rollback when rescheduling managed to improve occuoancy after
  remats.

lucas-rami committed:
== -> >= for achieved occupancy check at the end of the stage

This prevents rollbacking rematerializations in case rescheduling
improved occupancy even further than the target.

Also fix format.

lucas-rami committed:
Merge branch 'main' into pre-ra-remat-refactor

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,319 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    } else if (IncreaseOccupancy) {
+      // Check whether SGPR pressure prevents us from increasing occupancy.
+      if (ClearOptRegionsIf(NumSGPRs > MaxSGPRsMinOcc))
+        continue;
```
I think we can just return false if our DAG.MinOcc > OccBounds.first

lucas-rami replied:
I think we may want to keep going in case we encounter a region with ArchVGPR spilling after the current one.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,319 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+      if (hasExcessVGPRs(RP, MaxVGPRsIncOcc, ExcessRP, OccAGPRLimited)) {
+        // Check whether AGPR pressure prevents us from increasing occupancy.
+        if (ClearOptRegionsIf(OccAGPRLimited))
+          continue;
```
I think we can just return false if our DAG.MinOcc > OccBounds.first

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,319 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
-        HasRematDependency = true;
-        break;
-      }
+    if (NumAGPRs > MaxVGPRs) {
```
Can you set `ExcessArchVGPRs = NumArchVGPRs` -- we can spill AGPR to VGPR

lucas-rami replied:
Didn't know that, thanks.

jrbyrnes replied:
I think our best bet is to assume perfect allocation -- meaning we need NumAGPRs - MaxVGPRs number of available arch VGPR

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1675,319 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+  const unsigned MaxVGPRsNoSpill =
+      ST.getBaseMaxNumVGPRs(F, {ST.getMinNumVGPRs(OccBounds.second),
+                                ST.getMaxNumVGPRs(OccBounds.first)});
+  const unsigned MaxSGPRsMinOcc = ST.getMaxNumSGPRs(DAG.MinOccupancy, false);
```
Should this be DAG.MinOccupancy + 1?

lucas-rami replied:
Yep, thanks for the catch.

lucas-rami committed:
Address feedback

lucas-rami committed:
Merge branch 'main' into pre-ra-remat-refactor

lucas-rami committed:
Incorporate machine scheduler change from main

Now that single-MI regions are no longer filtered out by the machine
scheduler, we can greatly simplify the logic to update region boundaries
during rematerialization/rollbacking, at the cost of a single walk
through the MF's instructions at the beginning of the stage.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1682,333 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    }
+    return HasSpill;
+  }
+  if (RP.getVGPRNum(true) > MaxVGPRs) {
```
We should also take into account the limits of each of accvgpr and archvgpr.

For example, if we have 262 archgpr and 252 accvgpr, we have 514 vgpr (>= MaxVGPRs of 512 assuming occ lowerbound of 1, gfx942). 

This calculatuion determines ExcessArchVGPR = 262 - alignDown(512-252,4) = 2, but we really have ExcessArchVGPR of 6 due to archvgpr limit.

Something like this:

ExceedRF, ExceedArch, ExceedAcc -> ExcessArchVGPRs

{1,1,0} -> arch excess
{1,0,1} -> min (acc excess, arch usage)
{1,1,1} -> max (min (total excess, arch usage), min (acc excess, arch usage), arch excess)

lucas-rami replied:
Thanks for clarifying those edge cases for me, I was a bit unclear on some things and had not though of addressable register limits. I have made the excess RP tracking logic much more explicit (tracking excess above addressable limits and "regular" excess separately) through an `ExcessRP` internal helper struct, which should fix those issues and make it easier to integrate SGPR/AGPR remat in the future. The `ExcessRP` constructor basically replaces the `getExcessVGPRs` method.

lucas-rami committed:
Address feedback

- Allow remat in same block if using MI is in different region. Add unit
  test.
- Take into account excess ArchVGPR/AGPR pressure above addressable
  limits. Add unit tests to make sure this works.
- Fix unused register in existing unit test.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1682,333 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+      // We only care to rematerialize the instruction if it has a single
+      // non-debug user in a different block.
+      MachineInstr *UseMI = DAG.MRI.getOneNonDBGUser(Reg);
+      if (!UseMI || DefMI.getParent() == UseMI->getParent())
```
Yes we can easily do that now indeed. Rollbacking is "region-based" already so this doesn't require any change there.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1673,174 +1682,333 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+      ExcessArchVGPRs = NumArchVGPRs - alignDown(MaxVGPRs - NumAGPRs, 4);
+    }
+    return true;
+  }
```
Should be taken into account correctly now inside the `ExcessRP` constructor.

lucas-rami committed:
Merge branch 'main' into pre-ra-remat-refactor

Also update one recently changed unit test from main.

lucas-rami committed:
Remove manual recomputation of live intervals, use the LIS methods

This fails in some complex use cases, and it is not clear whether it is
meaningfully faster that just letting the LIS object recompute it from
scratch.

Also removed stale comment in isTriviallyRematerilizable: we can now
remat MIs with virtual register uses.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+      }
+      if ((Excess = ExcessRP(ST, RP, MaxVGPRsIncOcc))) {
+        // We can only rematerialize ArchVGPRs at this point.
+        unsigned NumArchVGPRsToRemat = Excess.ArchVGPRs + Excess.VGPRs;
```
should this be `std::max(Excess.ArchVGPRs, Excess.VGPRs)` ?

lucas-rami replied:
I think it should be the sum of the two. For non-unified RFs `Excess.VGPRs` is always 0 so it makes no difference. For unified RFs `Excess.ArchVGPRs` is only used to account for ArchVGPR pressure above the addressable limit, then the "regular excess pressure" is accounted for in Excess.VGPRs.

For example, on gfx942 with occ lower bound of 4, no AGPR usage and ArchVGPR usage of 257, we will get `MaxVGPRsIncOcc=512/4=128`, `Excess.ArchVGPRs=1`, `Excess.VGPRs=128`, and we will need to save 129 ArchVGPRs to hope to reach the min occupancy.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+}
+
+bool ExcessRP::saveArchVGPRs(unsigned NumRegs, bool UseArchVGPRForAGPRSpill) {
+  bool Progress = saveRegs(ArchVGPRs, NumRegs);
```
For many cases `NumRegs` will be 0 after the call to `saveRegs`. The only way I can make sense of the below handling is if `NumRegs` maintained it original value after the call to `saveRegs` -- for example, if our remat reduced VGPR pressure by 4, we should reduce both the `ExcessRP.ArchVGPRs` and `Excess.VGPRs` by this amount. Maybe `saveRegs` shouldn't take `NumRegs` by reference?

lucas-rami replied:
I think this is ok because `ExcessRP.ArchVGPRs` only tracks VGPR savings that must specifically be ArchVGPRs (because of the addressable limit). Once this has been reduced to 0 (e.g., on gfx942 once we have ArchVGPR usage <= 256) we start to use ArchVGPR savings as "generic VGPR savings".

Taking my above example again. On the first remat of an ArchVGPR-defining instruction (`NumRegs == 1`), `ArchVGPRs` will be reduced to 0 by this call and so will `NumRegs`, making the rest of the function void. On subsequent calls to `saveArchVGPRs`, this will change nothing to its arguments (so `NumRegs == 1` after the call) and return false, so we will hit `Progress |= saveRegs(VGPRs, NumRegs)` and start reducing the number of "generic" excess `VGPRs`.

jrbyrnes replied:
Oh I see -- you are not using the the ArchVGPR usage in calculating the VGPR usage, but are using instead the ArchVGPR target (i.e. MaxArchVGPRs).

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    // We may be able to save one more whole ArchVGPR allocation granule.
+    if (NumRegs >= ArchVGPRsToAlignment) {
+      NumSavedRegs += Granule;
+      ArchVGPRsToAlignment = Granule - (NumRegs - ArchVGPRsToAlignment);
```
I think `NumRegs - ArchVGPRsToAlignment` can be greater than `Granule`

lucas-rami replied:
From above unless I am mistaken we should get `NumRegs := NumRegs - (NumRegs / Granule) * Granule == NumRegs % Granule`, so `NumRegs < Granule`.

Additionally, we have `NumRegs >= ArchVGPRsToAlignment` from the if condition and `Granule >= ArchVGPRsToAlignment > 0` by construction. Hence,

$$NumRegs < Granule \Longrightarrow NumRegs - ArchVGPRsToAlignment < Granule - ArchVGPRsToAlignment < Granule$$

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+namespace {
+/// Models excess register pressure in a region and tracks our progress as we
+/// identify rematerialization opportunities.
+struct ExcessRP {
```
It would be nice if we could build better ExcessRP queries into GCNRegPressure and rewrite this heuristic to use those queries in combination with GCNRegPressure.inc() . This will help with code maintenance -- maybe as a followup PR

lucas-rami commented on llvm/test/CodeGen/AMDGPU/buffer-fat-pointers-memcpy.ll:
```diff
@@ -444,65 +444,77 @@ define amdgpu_kernel void @memcpy_known(ptr addrspace(7) inreg %src, ptr addrspa
-; GISEL-GFX942-NEXT:    v_accvgpr_write_b32 a4, v13 ; Reload Reuse
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[2:5], v62, s[4:7], 0 offen
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[6:9], v62, s[4:7], 0 offen offset:16
+; GISEL-GFX942-NEXT:    v_mov_b32_e32 v1, 0x2000
```
I checked the debug output. The upstream implementation identifies `v_mov_b32_e32 v1, 0x2000` as a candidate for rematerialization but eventually does not remat anything because it determines it won't be able to improve occupancy. My implementation remats the instruction to reduce identified spilling (4 VGPRs) in the region below (starting on line 448) by 1 VGPR. That's not enough to eliminate spilling but it reduces it.

jrbyrnes replied:
Okay -- I took a look at this test. The scheduler sees that after the rematerialization, we have improved the VGPR pressure which should theoretically help with spilling. However, the heuristic which decides if we are spilling or not is wrong, which I have mentioned in my other comment.

This is a bit unclear from just looking at the resultant code, because ,during RA, we are doing a suboptimal job and thus have spilling. This is due to how RA partitions the register file between AGPRs and VGPRs. We find our minimum allowable wavesPerEU is 4, so we have 128 vector registers, but RA splits this into 64 agprs and 64 vgprs. So, ironically, we have a limit of 64 archvgpr which is the same as the incorrect limit imposed by requiring an 8 occupancy during scheduuling. However, this is an accident and should be fixed as I think Matt is working on improving this. Also, if we use `attributes #0 = { "amdgpu-agpr-alloc"="0" }` we avoid the RA issue (and thus have no spiling), but the remat heuristics still think we are going to spill.

lucas-rami replied:
Thanks a lot for looking into this. With the spilling heuristic change you suggested the test now behaves as before.

jrbyrnes commented on llvm/test/CodeGen/AMDGPU/buffer-fat-pointers-memcpy.ll:
```diff
@@ -444,65 +444,77 @@ define amdgpu_kernel void @memcpy_known(ptr addrspace(7) inreg %src, ptr addrspa
 ; GISEL-GFX942-NEXT:    v_accvgpr_read_b32 v4, a1 ; Reload Reuse
 ; GISEL-GFX942-NEXT:    v_accvgpr_read_b32 v3, a2 ; Reload Reuse
 ; GISEL-GFX942-NEXT:    v_accvgpr_read_b32 v2, a3 ; Reload Reuse
+; GISEL-GFX942-NEXT:    v_cmp_lt_u32_e32 vcc, v0, v1
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[2:5], v62, s[4:7], 0 offen offset:32
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[14:17], v62, s[4:7], 0 offen offset:48
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[18:21], v62, s[4:7], 0 offen offset:64
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[22:25], v62, s[4:7], 0 offen offset:80
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[26:29], v62, s[4:7], 0 offen offset:96
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[30:33], v62, s[4:7], 0 offen offset:112
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[34:37], v62, s[4:7], 0 offen offset:128
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[38:41], v62, s[4:7], 0 offen offset:144
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[42:45], v62, s[4:7], 0 offen offset:160
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[46:49], v62, s[4:7], 0 offen offset:176
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[50:53], v62, s[4:7], 0 offen offset:192
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[54:57], v62, s[4:7], 0 offen offset:208
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
+; GISEL-GFX942-NEXT:    buffer_store_dwordx4 v[58:61], v62, s[4:7], 0 offen offset:224
+; GISEL-GFX942-NEXT:    s_waitcnt vmcnt(15)
```
waitcnts are needed for the results of the load, before we had a long waitcnt on line 479 so we had already waited on all the loads.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+  // groupsize induced occupancy bounds, producing unexpected results.
+  const unsigned MaxSGPRsNoSpill = ST.getBaseMaxNumSGPRs(
+      F, OccBounds, ST.getMaxNumPreloadedSGPRs(), ST.getReservedNumSGPRs(F));
+  const unsigned MaxVGPRsNoSpill =
```
I thought that `getOccupancyWithWorkGroupSizes` calculated the range of values for the upper bound of occupancy? Seems like what we need here is the minimum allowable occupancy.

For the changed test in buffer-fat-pointers-memcpy.ll we have `OccBounds` of [8,8]. Thus, we are saying that any pressure corresponding to an occupancy of less than 8 results in spilling. I don't see anything about this test that indicates we need an occupancy of at least 8 to avoid spilling (it makes more sense require at least 4 to fit the waves on the CU for the workgroup size 1024 case).

lucas-rami replied:
> I thought that getOccupancyWithWorkGroupSizes calculated the range of values for the upper bound of occupancy? Seems like what we need here is the minimum allowable occupancy.

Yes that's what `getOccupancyWithWorkGroupSizes` computes. Here I was assuming that, unless `amdgpu-waves-per-eu` is set, the minimum occupancy we want to allow is the mimimum induced by workgroup sizes. I see that other parts of the codebase don't use that same logic to determine a mimimum allowable occupancy, and so it's creating a disconnect that leads to suboptimal decisions in some situations as well as too aggressive remat in general.

I reverted this to using `AMDGPUSubtarget::getWavesPerEU`---which I still find somewhat wrong (for example with default workgroup sizes it will just ignore an override like `amdgpu-waves-per-eu="1,1"`)---but it should be more consistent with what other parts of the codebase use for that kind of consideration and gives 4 waves/EU as the minimum.

lucas-rami committed:
Revert to use of default waves/EU calculation & Fix progress detection

- Removes the custom logic for determining waves/EU bounds and use the
  default subtarget method.
- Saving ArchVGPRs without crossing an ArchVGPR granule is still
  progress.
- Some test churn.

lucas-rami commented:
Some pointers for the test churn:

- Generally our minimum allowable occupancy is now lower, which means that in some tests where the stage was trying to reduce spilling before it is now trying to increase occupancy. This leads to less remats in some cases (while still achieving the same occupancy as before) and a few instances where light code motion in tests is required to avoid rollbacks (which only happen when trying to increase occupancy) due to the excess pressure model being too optimistic. Adding the ability to continue remating if the heuristic was too optimistic seems like an important follow-up change.
- I changed the `omitted_subrange` test significantly because the "interesting MI" we want to be able to remat in it was no longer rematted with the new implementation; other MIs were evaluated first and were enough to improve occupancy. The test is now simpler and needs to remat this MI to be able to achieve higher occupancy.
- While fixing `reduce_arch_and_acc_vgrp_spill` I noticed some subtle issues in excess pressure tracking, notably in how progress is tracked for unified RFs when saving ArchVGPRs doesn't allow us to actually reduce pressure because of the ArchVGPR allocation granule. This is now resolved.
- Light changes in `reduce_spill_archvgpr_above_addressable_limit`/`reduce_spill_agpr_above_addressable_limit` are due to `AMDGPU::getWavesPerEU` ignoring `amdgpu-waves-per-eu` in some scenarios as I mention [here](https://github.com/llvm/llvm-project/pull/125885#discussion_r2033253965). The method considers that the minimum allowable occupancy is 4, when 1 should be ok per user request. I think this should be addressed separately, as it is really an issue of `AMDGPU::getWavesPerEU`.

jrbyrnes reviewed (approved):
This LGTM, but it would be good to wait for someone else to take a look due to the amount of change.

lucas-rami committed:
Get ArchVGPR alloc granule from AMDGPU::IsaInfo

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,420 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
-    if (!SRI->isVGPRClass(DAG.MRI.getRegClass(Reg)) ||
-        !DAG.MRI.hasOneDef(Reg) || !DAG.MRI.hasOneNonDBGUse(Reg))
-      continue;
+#ifndef NDEBUG
```
```suggestion
#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,420 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
-  DenseMap<unsigned, GCNRegPressure> NewPressure;
-  BitVector NewRescheduleRegions;
-  LiveIntervals *LIS = DAG.LIS;
+#ifndef DEBUG
```
```suggestion
#ifndef NDEBUG
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,420 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+    REMAT_DEBUG(dbgs() << "Occupancy minimal in regions:\n");
+  else
+    REMAT_DEBUG(dbgs() << "Spilling in regions:\n");
+  for (unsigned I = 0, E = DAG.Regions.size(); I != E; ++I)
+    if (auto OptIt = OptRegions.find(I); OptIt != OptRegions.end())
+      REMAT_DEBUG(dbgs() << "  " << I << ": " << OptIt->getSecond() << '\n');
```
```suggestion
  for (unsigned I = 0, E = DAG.Regions.size(); I != E; ++I) {
    if (auto OptIt = OptRegions.find(I); OptIt != OptRegions.end())
      REMAT_DEBUG(dbgs() << "  " << I << ": " << OptIt->getSecond() << '\n');
  }
```

lucas-rami committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,421 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+  const SIRegisterInfo *SRI = static_cast<const SIRegisterInfo *>(DAG.TRI);
+
+  REMAT_DEBUG(dbgs() << "Collecting rematerializable instructions in "
+                     << MF.getFunction().getName() << '\n');
```
printAsOperand to handle anonymous functions correctly

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,421 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
-  NewRescheduleRegions.resize(DAG.Regions.size());
+      // We only support rematerializing virtual VGPRs with one definition.
+      Register Reg = DefMI.getOperand(0).getReg();
+      if (!Reg.isVirtual() || !DAG.LIS->hasInterval(Reg) ||
```
Don't understand the hasInterval check - the verifier would fail if it were missing

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1696,421 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
 
-        assert(NewLiveIns[I].contains(UseReg));
-        LaneBitmask LiveInMask = NewLiveIns[I][UseReg];
+        assert(RegionLiveIns.contains(UseReg));
+        LaneBitmask LiveInMask = RegionLiveIns[UseReg];
```
Avoid double map lookup. Is this RegionLiveIns.at?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -2004,46 +2197,99 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+    return Region.first->getParent();
+  // The boundaries of an empty region may not point to a valid MI from which we
+  // can get the parent MBB, so we have to iterate over all the MF's MBBs.
+  for (MachineBasicBlock &MBB : MF)
+    if (MBB.end() == Region.second)
+      return &MBB;
```
```suggestion
  for (MachineBasicBlock &MBB : MF) {
    if (Region.second == MBB.end())
      return &MBB;
    }
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -2004,46 +2197,99 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+                                       RegionBoundaries &Region) {
+  if (Region.first != Region.second)
+    return Region.first->getParent();
+  // The boundaries of an empty region may not point to a valid MI from which we
+  // can get the parent MBB, so we have to iterate over all the MF's MBBs.
```
Shouldn't need to do this, the region should know what blocks it covers directly?

lucas-rami replied:
Just to clarify my rationale for this. When we end up emptying a region completely through rematerializations the region boundaries end up as `Region.first == Region.second` (`== BB.end()` if the region is the last in the BB). AFAIU I cannot retrieve the BB from the boundaries alone in this case, hence this linear search.

To avoid this linear search I added `PreRARematStage::RegionBB` which stores the parent MBB of each region before remats.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -2004,46 +2197,99 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+  if (!IncreaseOccupancy || MaxOcc >= TargetOcc)
+    return;
+
+  REMAT_DEBUG(dbgs() << "Rollbacking all rematerializations\n");
```
```suggestion
  REMAT_DEBUG(dbgs() << "Rolling back all rematerializations\n");
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -2004,46 +2197,99 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
+    return;
+
+  REMAT_DEBUG(dbgs() << "Rollbacking all rematerializations\n");
+  const auto *TII =
+      static_cast<const SIInstrInfo *>(MF.getSubtarget().getInstrInfo());
```
```suggestion
  const SIInstrInfo *TII = MF.getSubtarget<GCNSubtarget>().getInstrInfo();
```

lucas-rami committed:
Address review feedback

Aditionally fixes bug where MIRegion would only be filled after
collecting rematerializable instructions, however the collection phase
depends on MIRegion to determine whether the user of an MI is located in
a different region from that MI. This does not break any test.

lucas-rami committed:
[NFS] Directly use getMaxNum{S,V}GPRs methods

Using the base versions was necessary when we were not using the
default range of waves/EU. We use the default range now so we can
directly use the "regular" methods.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1688,174 +1703,421 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+}
+
+void PreRARematStage::rematerialize() {
+  const auto *TII =
+      static_cast<const SIInstrInfo *>(MF.getSubtarget().getInstrInfo());
```
```suggestion
  const SIInstrInfo *TII = MF.getSubtarget<GCNSubtarget>().getInstrInfo();
```

lucas-rami committed:
Update llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

jplehr commented:
I reverted this patch locally and that got rid of the issue.
Not sure if this can be fixed forward or needs a revert.

Pierre-vh commented:
It's just a few instructions that moved, I regenerated the test and committed it directly: 382a085a95b0abeac77b150b7b644b372bd08e78

jplehr commented:
Thank you

hubert-reinterpretcast commented:
@lucas-rami, are you working on a fix or should this be reverted?

lucas-rami commented:
Is this not the same issue solved by 382a085 and reported by a different builder that checked out LLVM at my commit (before the fix) each time? Line 509 of `mfma-loop.ll` on `main` is currently not the one reported by the CI output: `; GFX908-NEXT: s_mov_b32 s0, 16`

https://github.com/llvm/llvm-project/blob/d7987f1ce9bdc57fe10de6eef25fbe0df725c68f/llvm/test/CodeGen/AMDGPU/mfma-loop.ll#L509

`ninja-check-llvm-codegen-amdgpu` also reports no error when I build close to top-of-tree ([d7987f1](https://github.com/llvm/llvm-project/commit/d7987f1ce9bdc57fe10de6eef25fbe0df725c68f)).

Apologies if I am missing something here.

hubert-reinterpretcast commented:
> Is this not the same issue solved by [382a085](https://github.com/llvm/llvm-project/commit/382a085a95b0abeac77b150b7b644b372bd08e78) and reported by a different builder that checked out LLVM at my commit (before the fix) each time?

Sorry. I missed that the fix happened here: https://github.com/llvm/llvm-project/pull/125885#issuecomment-2862933136

vitalybuka commented:
There are still broken bots:
https://lab.llvm.org/buildbot/#/builders/52
https://lab.llvm.org/buildbot/#/builders/24/builds/8265
https://lab.llvm.org/buildbot/#/builders/169/builds/11275

slackito commented:
I get sanitizer errors when building at this commit with `-DLLVM_USE_SANITIZER=Address` and then running `ninja check-llvm-codegen-amdgpu`.

Here's the asan error:
```
==3331703==ERROR: AddressSanitizer: use-after-poison on address 0x7d39064ec8d4 at pc 0x55b90dc9ac25 bp 0x7ffff9c71550 sp 0x7ffff9c71548                                                                                            16:55:29 [69/2503]
READ of size 4 at 0x7d39064ec8d4 thread T0
    #0 0x55b90dc9ac24 in llvm::MachineInstr::getFlag(llvm::MachineInstr::MIFlag) const /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/CodeGen/MachineInstr.h:409:12
    #1 0x55b90dc9ac24 in llvm::MachineInstr::isBundledWithPred() const /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/CodeGen/MachineInstr.h:487:43
    #2 0x55b90dc9ac24 in llvm::SlotIndexes::removeMachineInstrFromMaps(llvm::MachineInstr&, bool) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/SlotIndexes.cpp:129:3
    #3 0x55b909936baa in llvm::LiveIntervals::RemoveMachineInstrFromMaps(llvm::MachineInstr&) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/CodeGen/LiveIntervals.h:296:14
    #4 0x55b909936baa in llvm::PreRARematStage::rematerialize() /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:2090:14
    #5 0x55b9099314ef in llvm::PreRARematStage::initGCNSchedStage() /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:1117:3
    #6 0x55b90992e90a in llvm::GCNScheduleDAGMILive::runSchedStages() /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:966:17
    #7 0x55b90d79d811 in llvm::impl_detail::MachineSchedulerBase::scheduleRegions(llvm::ScheduleDAGInstrs&, bool) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/MachineScheduler.cpp:825:13
    #8 0x55b90d79b8f3 in llvm::impl_detail::MachineSchedulerImpl::run(llvm::MachineFunction&, llvm::TargetMachine const&, llvm::impl_detail::MachineSchedulerImpl::RequiredAnalyses const&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/
MachineScheduler.cpp:484:3
    #9 0x55b90d7cdd43 in (anonymous namespace)::MachineSchedulerLegacy::runOnMachineFunction(llvm::MachineFunction&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/MachineScheduler.cpp:576:15
    #10 0x55b90d5d5377 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/MachineFunctionPass.cpp:108:10
    #11 0x55b90e2d524a in llvm::FPPassManager::runOnFunction(llvm::Function&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:1406:27
    #12 0x55b90c933051 in (anonymous namespace)::CGPassManager::RunPassOnSCC(llvm::Pass*, llvm::CallGraphSCC&, llvm::CallGraph&, bool&, bool&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:180:25
    #13 0x55b90c933051 in (anonymous namespace)::CGPassManager::RunAllPassesOnSCC(llvm::CallGraphSCC&, llvm::CallGraph&, bool&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:469:9
    #14 0x55b90c933051 in (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:534:18
    #15 0x55b90e2d61be in (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:1521:27
    #16 0x55b90e2d61be in llvm::legacy::PassManagerImpl::run(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:539:44
    #17 0x55b90870ff3e in compileModule(char**, llvm::LLVMContext&) /usr/local/google/home/jgorbe/code/llvm/llvm/tools/llc/llc.cpp:753:8
    #18 0x55b90870af9e in main /usr/local/google/home/jgorbe/code/llvm/llvm/tools/llc/llc.cpp:400:22
    #19 0x7f2907033ca7 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #20 0x7f2907033d64 in __libc_start_main csu/../csu/libc-start.c:360:3
    #21 0x55b908622930 in _start (/dev/shm/jgorbe-code/lldb-build/bin/llc+0x905b930)

0x7d39064ec8d4 is located 4052 bytes inside of 4096-byte region [0x7d39064eb900,0x7d39064ec900)
allocated by thread T0 here:
    #0 0x55b908709712 in operator new(unsigned long, std::align_val_t, std::nothrow_t const&) /usr/local/google/home/jgorbe/forks/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:104:3
    #1 0x55b90faa24af in llvm::allocate_buffer(unsigned long, unsigned long) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Support/MemAlloc.cpp:16:18
    #2 0x55b908bef425 in llvm::MallocAllocator::Allocate(unsigned long, unsigned long) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/Support/AllocatorBase.h:92:12
    #3 0x55b908bef425 in llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>::StartNewSlab() /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/Support/Allocator.h:346:42
    #4 0x55b908bef1a9 in llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>::AllocateSlow(unsigned long, unsigned long, llvm::Align) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/Support/Allocator.h:202:5
    #5 0x55b90d5b6625 in llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>::Allocate(unsigned long, unsigned long) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/Support/Allocator.h:216:12
    #6 0x55b90d5b6625 in llvm::MachineInstr* llvm::Recycler<llvm::MachineInstr, 80ul, 8ul>::Allocate<llvm::MachineInstr, llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>>(llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator,
4096ul, 4096ul, 128ul>&) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/Support/Recycler.h:95:57
    #7 0x55b90d5b6625 in llvm::MachineFunction::CreateMachineInstr(llvm::MCInstrDesc const&, llvm::DebugLoc, bool) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/MachineFunction.cpp:429:35
    #8 0x55b908ca3454 in llvm::BuildMI(llvm::MachineBasicBlock&, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, llvm::MIMetadata const&, llvm::MCInstrDesc const&, llvm::Register) /usr/local/google/home/jgorbe/code/llvm/llvm/include
/llvm/CodeGen/MachineInstrBuilder.h:397:25
    #9 0x55b90f3f0a78 in llvm::BuildMI(llvm::MachineBasicBlock*, llvm::MIMetadata const&, llvm::MCInstrDesc const&, llvm::Register) /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/CodeGen/MachineInstrBuilder.h:498:10
    #10 0x55b90f3f0a78 in llvm::FunctionLoweringInfo::set(llvm::Function const&, llvm::MachineFunction&, llvm::SelectionDAG*) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp:307:11
    #11 0x55b90f6afb74 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:579:13
    #12 0x55b9094a470a in llvm::AMDGPUDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:128:28
    #13 0x55b90f6a9adf in llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:374:20
    #14 0x55b90d5d5377 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/CodeGen/MachineFunctionPass.cpp:108:10
    #15 0x55b90e2d524a in llvm::FPPassManager::runOnFunction(llvm::Function&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:1406:27
    #16 0x55b90c933051 in (anonymous namespace)::CGPassManager::RunPassOnSCC(llvm::Pass*, llvm::CallGraphSCC&, llvm::CallGraph&, bool&, bool&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:180:25
    #17 0x55b90c933051 in (anonymous namespace)::CGPassManager::RunAllPassesOnSCC(llvm::CallGraphSCC&, llvm::CallGraph&, bool&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:469:9
    #18 0x55b90c933051 in (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/Analysis/CallGraphSCCPass.cpp:534:18
    #19 0x55b90e2d61be in (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:1521:27
    #20 0x55b90e2d61be in llvm::legacy::PassManagerImpl::run(llvm::Module&) /usr/local/google/home/jgorbe/code/llvm/llvm/lib/IR/LegacyPassManager.cpp:539:44
    #21 0x55b90870ff3e in compileModule(char**, llvm::LLVMContext&) /usr/local/google/home/jgorbe/code/llvm/llvm/tools/llc/llc.cpp:753:8
    #22 0x55b90870af9e in main /usr/local/google/home/jgorbe/code/llvm/llvm/tools/llc/llc.cpp:400:22
    #23 0x7f2907033ca7 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

SUMMARY: AddressSanitizer: use-after-poison /usr/local/google/home/jgorbe/code/llvm/llvm/include/llvm/CodeGen/MachineInstr.h:409:12 in llvm::MachineInstr::getFlag(llvm::MachineInstr::MIFlag) const
Shadow bytes around the buggy address:
  0x7d39064ec600: f7 00 00 00 00 00 00 00 00 00 00 f7 00 00 00 00
  0x7d39064ec680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7d39064ec700: 00 00 00 00 00 00 00 00 00 00 00 00 f7 00 00 00
  0x7d39064ec780: 00 00 00 00 00 00 00 f7 00 00 00 00 00 00 00 00
  0x7d39064ec800: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 00 00 00 00
=>0x7d39064ec880: 00 00 00 00 f7 f7 f7 f7 f7 f7[f7]f7 f7 f7 f7 f7
  0x7d39064ec900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7d39064ec980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7d39064eca00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7d39064eca80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7d39064ecb00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3331703==ABORTING
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /dev/shm/jgorbe-code/lldb-build/bin/FileCheck -check-prefix=GCN -check-prefix=SI /dev/shm/jgorbe-code/llvm/llvm/test/CodeGen/AMDGPU/vgpr-spill-emergency-stack-slot.ll
```

vitalybuka commented:
Yes, the same on bots.

vitalybuka commented:
@lucas-rami we had to revert to fix bots

lucas-rami commented:
Sounds good, thanks for the ping and sorry for the faulty commit, working on addressing the issue.
